src/mito2/src/read/merge.rs

```diff
@@ -201,32 +219,143 @@ impl BatchMerger {
 
     /// Merge all buffered batches and returns the merged batch. Then
     /// reset the buffer.
-    fn merge_batches(&mut self) -> Result<Option<Batch>> {
+    fn merge_batches(&mut self) -> Result<VecDeque<Batch>> {
         if self.batches.is_empty() {
-            return Ok(None);
+            return Ok(VecDeque::new());
         }
 
-        let batches = mem::take(&mut self.batches);
-        // Concat all batches.
-        let mut batch = Batch::concat(batches)?;
-
-        // TODO(yingwen): metrics for sorted and unsorted batches.
-        if !self.is_sorted {
-            // Slow path. We need to merge overlapping batches. For simplicity, we
-            // just sort the all batches and remove duplications.
-            batch.sort_and_dedup()?;
```

Comment:
Question: can the `sort_and_dedup` have a better performance if all batches are small?

Reply:
Maybe. Concatenating and sorting batches are costly so I guess `sort_and_dedup()` might outperform merging by heap only when the total number of rows is small enough. However, we need some benchmarks to prove this and the cost between the two solutions in this situation should be close. We can try this optimization later if this is really a common case.